### PR TITLE
[DNM] net: allow relocation of network subsystem to RAM

### PR DESCRIPTION
**Pull request by danieldegrasse**
Enable relocation of network subsystem to RAM. The intent of this feature is to enable SOCs with fast instruction TCM to take advantage of the improved performance available by relocating networking code to this region. This PR enables this feature for the Cortex M7 based iMX RT SOCs with a Kinetis Ethernet controller.

This PR is dependent on the following PRs (and currently integrates the commits from both):
#50791
#50790

Below are some TCP performance results using this PR. Results were obtained using iperf 2.1.5 with the zperf sample running on the Zephyr target. These results relocate the networking stack, L2 ethernet implementation, and NXP ethernet driver to ITCM.

| Board | SHA | TCP RX | TCP TX |
| ----- | ------ | ------- | ------- |
| mimxrt1060_evk | main (fd9e5429a37f24780f5e27c08ecb470d5d36e93a) | 44.2Mbps | 43.7Mbps |
| mimxrt1060_evk | This PR | 94.5Mbps | 71.6Mbps |
| mimxrt1050_evk | main (fd9e5429a37f24780f5e27c08ecb470d5d36e93a) | 60.2Mbps | 68.5Mbps |
| mimxrt1050_evk | This PR | 94.5Mbps | 71.4Mbps|

### CODE REVIEW

**Comment by jukkar** [comment]
Shouldn't this be a board feature and not a networking one? I am asking because you introduce new Kconfig symbols to networking stack which imho do not seem to be really needed there.

**Comment by jukkar** [comment]
The feature itself is nice if we get such a speed increase with minimal changes. Just pondering where / how this should be setup.

**Comment by tbursztyka** [comment]
Looks fine to me to get a generic config option on net stack, that some hw can take advantage of. It's actually cleaner that way I think. Many more SoC/boards are going to use that at some point

File: subsys/net/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,10 @@ zephyr_library()
 zephyr_library_sources_ifdef(CONFIG_NET_BUF             buf.c)
 zephyr_library_sources_ifdef(CONFIG_NET_HOSTNAME_ENABLE hostname.c)
 
+if(CONFIG_NET_CODE_RELOCATE)
```
Comment (tbursztyka) [inline subsys/net/CMakeLists.txt:7]: commit title would to be revised, you are relocating both the net stack and the ethernet L2. And there is no dependency with each other on this. Perhaps splitting the patch in 2 would be clearer: first the net stack, then the ethernet L2.

**Comment by jukkar** [comment]
@danieldegrasse I have no strong opinion about this, I was just thinking aloud and pondering about how to setup this. Overall, this is a nice speed improvement for this soc.

**Review by tbursztyka — APPROVED** [review]

**Pull request closed without merging**